### PR TITLE
Curlib optimizations

### PR DESCRIPTION
**Pull request by szsascha**
#146 

### Checklist

* [x] have tested my change
* [x] updated relevant documentation
* [x] Remove any/all `console.log`s I added
* [x] eslint is not complaining
* [x] have added myself to the contributors' list in the README

### CODE REVIEW

**Comment by worksofliam** [comment]
@szsascha Thanks for this PR.

I updated the original comment to include the required checklist. After those items are complete and I've had a little test, it can be merged.

Thanks!

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,6 @@
 node_modules
 .vscode-test/
 *.vsix
-dist
\ No newline at end of file
+dist
+package-lock.json
```
Comment (worksofliam) [inline .gitignore:5]: `package-lock.json` must not be in the .gitignore.

File: src/api/IBMi.js [hunk 2]
```diff
@@ -90,7 +90,7 @@ module.exports = class IBMi {
           undefined,
         );
       } catch (e) {
-        let [errorcode, errortext] = e.split(`:`);
+        let [errorcode] = e.split(`:`)[0];
```
Comment (worksofliam) [inline src/api/IBMi.js:93]: What was the reason for this change? I suppose ESLint was just saying it was undefined?

For the moment, please revert this change as I am about to start using the `errorText` in another PR I'm in.

Comment (szsascha) [reply src/api/IBMi.js]: I just didn't want any open warnings :)

But I've reverted the commit.

**Comment by worksofliam** [comment]
@szsascha I am going to test this now.

**Comment by worksofliam** [comment]
Okay, so I remember why we didn't add `CUR` or `SYS` libraries to the `defaultUserLibraries` prop - that's because we can't remove them from the library list at compile time.

We do this at compile time:

```js
              commandResult = await connection.qshCommand([
                `liblist -d ` + connection.defaultUserLibraries.join(` `),
                `liblist -a ` + libl.join(` `),
                command,
              ], undefined, 1);
```

We remove the default library list (since each compile runs in its own job) and then set it to the user's library list in the extension.

Maybe, instead of this change, we actually need to display the `CUR` and `SYSTEM` libraries in the library list view, but not let the user move/remove them.

What do you think @szsascha?

**Comment by worksofliam** [comment]
Next big crazy idea:

We completely replace `&BUILDLIB` with `&CURLIB` and run `liblist -c ${lib}` at compile time to set the current library.

**Comment by szsascha** [comment]
@worksofliam 

Sorry, but I can't really form an opinion on this at the moment. For this I lack a bit of background on how the whole plugin is working.

I would assume that it makes the most sense to have the library list the same like you would have it when you compile it on i. This includes the CUR and SYS libraries. You need this for some cases. For example when you are using an object from QSYS without specifying a library.

I also had a look on the qsh liblist command documentation (https://www.ibm.com/docs/en/i/7.4?topic=jobs-liblist)
It says: "Remove library from the user portion of the library list or **unset the current library if the -c option is also specified.**"
So I guess it's possible to remove it?
The SYS part should be set automatically by the system.

How do we want to continue?

**Comment by szsascha** [comment]
> We completely replace &BUILDLIB with &CURLIB and run liblist -c ${lib} at compile time to set the current library.

Sounds like a plan! Do we then still need &BUILDLIB for anything?

**Comment by worksofliam** [comment]
@szsascha Let's go with adding `&CURLIB`. In fact, I can do this change for us.

I think we will keep `&BUILDLIB` around for backwards compatibility (since some Actions might use them), but it will be the same as `&CURLIB`.

**Comment by szsascha** [comment]
> When I try and compile on a new connection with this change:
>
> Library list: QSYS2 QGPL QTEMP
> Command: CRTBNDRPG PGM(BARRY/BETH) SRCFILE(BARRY/QRPGLESRC) OPTION(*EVENTF) DBGVIEW(*SOURCE) TGTRLS(*CURRENT)
> liblist: 001-0083 Error CPF2104 found deleting library QSYS2 from library list.
> liblist: 001-0081 Error CPF2103 found adding library QSYS2 to library list.

I think we can completely ignore the SYS list at this point. The system will handle it by itself. You're also not able to change it with EDTLIBL.

**Comment by szsascha** [comment]
> @szsascha Let's go with adding &CURLIB. In fact, I can do this change for us.

@worksofliam Ok, thank you! I already have an idea for my next contribution :)

**Comment by worksofliam** [comment]
@szsascha Awesome - I can't wait to see it! My PR will be ready within the next 30 minutes or so. I will ping you so you can see it.

I also want to mention that I am sorry about how confusing this initial PR was. The app is quite big now, and playing with the library list affects the core piece of the extension (e.g. the compliation step).

Hope to make it easier in the future.
